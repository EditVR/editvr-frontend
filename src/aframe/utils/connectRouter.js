/**
 * @file connectRouter.js
 * Utility that assists in connecting AFrame components to React Router.
 */

import { matchPath } from 'react-router';
import history from '../../lib/routerHistory';

/**
 * Attaches router state to a given component's schema.
 *
 * @param {object} aframeComponent - AFrame component object.
 * @param {string} pathPattern - React Router path pattern that is expected.
 *
 * @returns {object} - AFrame component object with router state.
 */
const connect = (aframeComponent, pathPattern = null) => {
  const component = aframeComponent;

  console.log(history);

  // Default component schema to an empty object.
  component.schema = component.schema || {};

  component.schema = {
    ...component.schema,
    router: {
      history,
      match: pathPattern
        ? matchPath(history.location.pathname, {
            path: pathPattern,
            exact: pathPattern.includes('?')
          })
        : null
    }
  };

  return component;
};

export default connect;
