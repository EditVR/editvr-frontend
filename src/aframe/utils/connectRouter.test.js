/**
 * @file connectRouter.test.js
 * Contains tests for connectRouter.
 */

import connectRouter from './connectRouter';

describe('aframe->utils->connectRouter()', () => {
  it('Can connect an AFrame component to React Router.', () => {
    const init = jest.fn();
    const didReceiveRoute = jest.fn();
    const shouldComponentUpdateRouting = jest.fn(() => true);
    const connected = connectRouter(
      {
        init,
        didReceiveRoute,
        shouldComponentUpdateRouting
      },
      '/experience/vreditor/:experienceSlug/:sceneSlug',
      {
        listen: jest.fn(() => jest.fn()),
        location: {
          pathname: '/experience/vreditor/experience/scene'
        }
      }
    );

    // Execute the init handler, as AFrame would.
    connected.init();
    expect(init).toHaveBeenCalledTimes(1);
    expect(connected).toMatchSnapshot();

    // Run router change handlers.
    connected.handleRouterChange.call(connected);
    expect(connected.shouldComponentUpdateRouting).toHaveBeenCalledTimes(1);
    expect(didReceiveRoute).toHaveBeenCalledTimes(1);
  });

  it('Does remove its subscription to router history when destructed..', () => {
    const remove = jest.fn();
    const unsubscribe = jest.fn();
    const listen = jest.fn(() => unsubscribe);
    const connected = connectRouter(
      { remove },
      '/experience/vreditor/:experienceSlug/:sceneSlug',
      {
        listen,
        location: {
          pathname: '/experience/vreditor/experience/scene'
        }
      }
    );

    // Execute the init handler, as AFrame would.
    connected.init();

    // Un-mount component and note that the subscription to router is destructed.
    connected.remove();
    expect(remove).toHaveBeenCalledTimes(1);
    expect(unsubscribe).toHaveBeenCalledTimes(1);
  });
});
