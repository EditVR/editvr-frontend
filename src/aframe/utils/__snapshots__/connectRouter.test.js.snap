// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aframe->utils->connectRouter() Can connect an AFrame component to React Router. 1`] = `
Object {
  "didReceiveProps": [Function],
  "handleRouterChange": [Function],
  "init": [Function],
  "remove": [Function],
  "router": Object {
    "history": Object {
      "listen": [Function],
      "location": Object {
        "pathname": "/experience/vreditor/experience/scene",
      },
    },
    "match": Object {
      "isExact": true,
      "params": Object {
        "experienceSlug": "experience",
        "sceneSlug": "scene",
      },
      "path": "/experience/vreditor/:experienceSlug/:sceneSlug",
      "url": "/experience/vreditor/experience/scene",
    },
  },
  "shouldComponentUpdateRouting": [Function],
  "unsubscribeFromRouter": [Function],
}
`;
